Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -8,6 +8,63 @@ For all of the test cases below, it need to be investigated first what kind of a
+6. If the access token is valid, the resource server (API) serves the resource to the application
+
+Source: https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2
+
+The authorization grant might have different forms (explicit, implicit, etc), the access token might have a companion refresh token and might be used on different ways (e.g., as a bearer token). Please refer to https://tools.ietf.org/html/rfc6749 for further details.
+
+These are some of the common best practices for OAuth2 on native apps:
+
+User-agent:
+- Use an external user-agent (the browser) instead of an embedded user-agent (e.g. WebView or internal client user interface) to prevent End-User Credentials Phishing. However, doing this the app relies on the OS Keychain for server trust. This way it will not be possible to implement certificate pinning. A solution for this would be to restrict the embedded user-agent to only the relevant domain.
+- The user should have a way to verify visual trust mechanisms (e.g., Transport Layer Security (TLS) confirmation, web site mechanisms).
+- The client should validate the fully qualified domain name of the server to the public key presented by the server during connection establishment to prevent man-in-the-middle attacks.
+
+Type of grant:
+- Use code grant instead of implicit grant on native apps.
+- When using code grant, implement PKCE (Proof Key for Code Exchange) to protect the code grant. Make sure that the server also implements it.
+- The auth "code" should short-lived and only used inmidiately after receiving it. Make sure that they only reside on transient memory and are not stored or logged.
+
+Client secrets:
+- No shared secret should be used as proof of the client's identity as this could lead to client impersonation ("client_id" already serves this purpose). If for some reason they do use client secrets, be sure that they are stored in secure local storage.
+
+End-User credentials
+- The transmission of end-user credentials must be protected using transport-layer mechanisms such as TLS or any alternative that avoids the sending of plaintext credentials over the wire (e.g., Hash-based Message Authentication Code).
+
+Tokens:
+- Keep access tokens in transient memory.
+- Access tokens must be securely transmitted using secure transport (TLS).
+- The scope and expiry time of access tokens should be reduced when end-to-end confidentiality cannot be guaranteed or when the token provides access to high risk information or action.
+- Remember that if the app uses access tokens as bearer tokens and no additional mechanism is used to identify the client, the attacker can access all resources associated with the token and its scope.
+- Store refresh tokens in secure local storage as they are long-term credentials.
+
+For additional best practices and detailed information please refer to the source documents:
```

Comment:
I would add a more detailed subchapter of common insecure implementations here, e.g. lack of verification that user-id is bound to the received access token. More info is available on the recent BH writeup here: https://www.blackhat.com/docs/eu-16/materials/eu-16-Yang-Signing-Into-Billion-Mobile-Apps-Effortlessly-With-OAuth20-wp.pdf